### PR TITLE
feat: Container resource requirements and HPA

### LINKED ISSUE #27: Container resource requirements and HPA
### Description of the feature or enhancement:
Containers can set their resource limits/requests.

### Use Case:
Horizontal Pod Autoscaler requirs containers to set resource requirements. 

### Proposed Solution:
```ts
const container = deployment.addContainer({
 resources: {
   cpu: {
     request: Cpu.millis(100),
     limit: Cpu.units(0.5),
   }
 }
});
container.resources.memory.request = Size.mebibytes(100);
container.resources.memory.limit = Size.gibibytes(1);
// optional? 
container.resources.setHugepage("2Mi", {
  ...
})
```

### Other:
* [x] :wave: I may be able to implement this feature request
* [x] :warning: This feature might incur a breaking change

---

This is a :rocket: Feature Request

### PR DESCRIPTION
**Pull request by mac-atre**
Allow setting compute resources (CPU and memory requests and limits) required by the container.

Signed-off-by: Mac Atre <mac.atre@gmail.com>

Fixes #27

### CODE REVIEW

File: src/container.ts [hunk 1]
```diff
@@ -536,6 +551,38 @@ export class Container {
       });
     }
 
+    // Resource requests and limits
```
Comment (iliapolo) [inline src/container.ts:554]: This feels a little cluttered...can you try making the flow clearer? maybe something like:

```ts
    const cpuLimit = this.resources?.cpu?.limit?.amount;
    const cpuRequest = this.resources?.cpu?.request?.amount;
    const memoryLimit = this.resources?.memory?.limit;
    const memoryRequest = this.resources?.memory?.request;

    const limits: { [key: string]: k8s.Quantity } = {};
    const requests: { [key: string]: k8s.Quantity } = {};

    if (cpuLimit) {
      limits.cpu = k8s.Quantity.fromString(cpuLimit);
    }
    if (memoryLimit) {
      limits.memory = k8s.Quantity.fromString(memoryLimit.toMebibytes().toString() + 'Mi');
    }
    if (cpuRequest) {
      requests.cpu = k8s.Quantity.fromString(cpuRequest);
    }
    if (memoryRequest) {
      requests.memory = k8s.Quantity.fromString(memoryRequest.toMebibytes().toString() + 'Mi');
    }

    const resourceRequirements = {
      limits: limits,
      requests: requests,
    };
```

?

File: src/container.ts [hunk 2]
```diff
@@ -662,6 +710,47 @@ export enum MountPropagation {
   BIDIRECTIONAL = 'Bidirectional',
 }
 
+/**
+ * CPU and memory compute resources
+ */
+export interface IResources {
```
Comment (iliapolo) [inline src/container.ts]: We normally use `IXxx` naming for [imported](https://github.com/cdk8s-team/cdk8s-plus/blob/k8s-22/main/src/secret.ts#L24) resources, not for inputs. 

Just remove the `I` prefix.

**Review by iliapolo — CHANGES_REQUESTED** [review]

File: src/container.ts (hunk 1 above)
Comment (mac-atre) [reply src/container.ts:554]: Thanks @iliapolo. I just committed the changes above except for the following block:
```
    let resourceRequirements: ResourceRequirements | undefined = undefined;
    if (Object.keys(limits).length > 0 || Object.keys(requests).length > 0) {
      resourceRequirements = {
        limits: limits,
        requests: requests,
      };
    }
```
If I don't allow `resourceRequirements` to be undefined (when neither limits nor requests are set), it ends up being an empty object, and then the 'Instantiation properties are all respected' test fails.

**Review by iliapolo — APPROVED** [review]

File: src/container.ts (hunk 1 above)
Comment (iliapolo) [reply src/container.ts:554]: Fair enough 👍

**Pull request merged**